File: finalFacebook/data strcuture/Closest Binary Search Tree Value.java
// dfs O(N) O(N) 
class Solution {
    public int closestValue(TreeNode root, double target) {
        Stack<TreeNode> stack = new Stack<>();
        int res = 0;
        double diff = -1.0;
        stack.push(root);
         while(!stack.isEmpty()) {
             TreeNode node = stack.pop();
             double temp = (double) node.val;
             double newDiff = Math.abs(target - temp);
             if(diff == -1.0 || diff > newDiff) {
                 diff = newDiff;
                 res = node.val;
             }
             if(node.left != null && node.val > target) {
                 stack.push(node.left);
             }
             else if(node.right != null && node.val < target) {
                 stack.push(node.right);
             }
         }
        return res;
    }
}

File: Python/Tree/Closest Binary Search Tree Value.py
#   O(K)    O(H)
class Solution:
    def closestValue(self, root: TreeNode, target: float) -> int:
        stack, pred = [], float('-inf')
        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                if pred <= target and target < root.val:
                    return min(pred, root.val, key = lambda x: abs(target - x))    
                pred = root.val
                root = root.right
        return pred

class Solution:
    def closestValue(self, root: TreeNode, target: float) -> int:
        closest = root.val
        while root:
            closest = min(root.val, closest, key = lambda x: abs(target - x))
            root = root.left if target < root.val else root.right
        return closest